pkg/acsengine/orchestrators.go

```diff
@@ -0,0 +1,273 @@
+func GetOrchestratorUpgradeInfo(cs *api.ContainerService) (*api.OrchestratorInfo, error) {
+	if cs == nil || cs.Properties == nil || cs.Properties.OrchestratorProfile == nil {
+		return nil, fmt.Errorf("Incomplete ContainerService")
+	}
+	if len(cs.Properties.OrchestratorProfile.OrchestratorRelease) == 0 {
+		return nil, fmt.Errorf("Missing Orchestrator Release")
+	}
+	if cs.Properties.OrchestratorProfile.OrchestratorType != api.Kubernetes {
+		return nil, fmt.Errorf("Upgrade operation is not supported for '%s'", cs.Properties.OrchestratorProfile.OrchestratorType)
+	}
+	arr, err := kubernetesInfo(&api.VersionInfo{
+		Release: cs.Properties.OrchestratorProfile.OrchestratorRelease,
+		Version: cs.Properties.OrchestratorProfile.OrchestratorVersion})
+	if err != nil {
+		return nil, err
+	}
+	// has to be exactly one element per specified orchestrator/release
+	if len(arr) != 1 {
+		return nil, fmt.Errorf("Umbiguous Orchestrator Releases")
+	}
+	return arr[0], nil
+}
+
+func kubernetesInfo(csInfo *api.VersionInfo) ([]*api.OrchestratorInfo, error) {
+	orchs := []*api.OrchestratorInfo{}
+	if len(csInfo.Release) == 0 {
+		// get info for all supported versions
+		for rel, ver := range common.KubeReleaseToVersion {
+			upgrades, err := kubernetesUpgrades(&api.VersionInfo{Release: rel, Version: ver})
+			if err != nil {
+				return nil, err
+			}
+			orchs = append(orchs,
+				&api.OrchestratorInfo{
+					Orchestrator: api.Kubernetes,
+					VersionInfo: api.VersionInfo{
+						Release: rel,
+						Version: ver,
+					},
+					Default:    rel == common.KubernetesDefaultRelease,
+					Upgradable: upgrades,
+				})
+		}
+	} else {
+		// get info for the specified release
+		ver, ok := common.KubeReleaseToVersion[csInfo.Release]
+		if !ok {
+			return nil, fmt.Errorf("Kubernetes release %s is not supported", csInfo.Release)
+		}
+		// set defaulr version if empty
+		if len(csInfo.Version) == 0 {
+			csInfo.Version = ver
+		}
+		upgrades, err := kubernetesUpgrades(csInfo)
+		if err != nil {
+			return nil, err
+		}
+		orchs = append(orchs,
+			&api.OrchestratorInfo{
+				Orchestrator: api.Kubernetes,
+				VersionInfo: api.VersionInfo{
+					Release: csInfo.Release,
+					Version: ver,
+				},
+				Default:    csInfo.Release == common.KubernetesDefaultRelease,
+				Upgradable: upgrades,
+			})
+	}
+	return orchs, nil
+}
+
+func kubernetesUpgrades(csInfo *api.VersionInfo) ([]*api.VersionInfo, error) {
+	ret := []*api.VersionInfo{}
+	var csVer, pVer versionNumber
+
+	if err := csVer.parse(csInfo.Version); err != nil {
+		return ret, err
+	}
+	switch csInfo.Release {
+	case common.KubernetesRelease1Dot5:
+		// add next release
+		ret = append(ret, &api.VersionInfo{Release: common.KubernetesRelease1Dot6, Version: common.KubeReleaseToVersion[common.KubernetesRelease1Dot6]})
+	case common.KubernetesRelease1Dot6:
+		// check for patch upgrade
+		patchVersion := common.KubeReleaseToVersion[common.KubernetesRelease1Dot6]
```

Comment:
Looks like a lot of code repetition within each version. Can be extracted and reused?